lib/galaxy/workflow/resources/__init__.py

```diff
@@ -0,0 +1,176 @@
+            user_groups.append(g.group.name)
+        default_group = by_group.get('default', None)
+        for group_name, group_def in by_group.get("groups", {}).items():
+            if group_name == default_group or group_name in user_groups:
+                for tag in group_def:
+                    if type(tag) is dict:
+                        if tag.get('name') not in user_permissions:
+                            user_permissions[tag.get('name')] = {}
+                        for option in tag.get('options'):
+                            user_permissions[tag.get('name')][option] = {}
+                    else:
+                        if tag not in user_permissions:
+                            user_permissions[tag] = {}
+
+        # user_permissions is now set.
+        params = get_workflow_parameter_list(workflow_resource_params, user_permissions)
+    return params
+
+
+# returns an array of parameters that a users set of permissions can access.
+def get_workflow_parameter_list(params, user_permissions):
+    param_list = []
+    for param_name, param_elem in params.items():
+        attr = deepcopy(param_elem.attrib)
+        if attr['name'] in user_permissions:
+            # Allow 'select' type parameters to be used
+            if attr['type'] == 'select':
+                option_data = []
+                reject_list = []
+                for option_elem in param_elem.findall("option"):
+                    if option_elem.attrib['value'] in user_permissions[attr['name']]:
+                        option_data.append({
+                            'label': option_elem.attrib['label'],
+                            'value': option_elem.attrib['value']
+                        })
+                    else:
+                        reject_list.append(option_elem.attrib['label'])
+                attr['data'] = option_data
+                attr_help = ""
+                if 'help' in attr:
+                    attr_help = attr['help']
+                if reject_list:
+                    attr_help += "<br/><br/>The following options are available but disabled.<br/>" + \
+                                 str(reject_list) + \
+                                 "<br/>If you believe this is a mistake, please contact your Galaxy admin."
+                attr['help'] = attr_help
+
+            param_list.append(attr)
+    return param_list
+
+
+def validate_by_group_workflow_parameters_mapper(by_group, workflow_resource_params):
+    # TODO: Rework this to consume YAML mapper file, drop all parameter validation I think?
+
+    # TODO: Do we really want to remove all parameter validation? I have changed it to validate the yml against the
+    # workflow_resource_params so that we disallow selection when the yml is not correct since that file would be
+    # changed the most frequently and most prone to configuration errors.
+
+    # TODO: finish validation? Currently validating needed stuff except for 'options' in a 'select'
```

Comment:
This validation is fine - it only applies to the by_group variant of this - which I think is what I was worried about. If ultimately "by_group" could look a bit more like a plugin and not be dispatched in ifs so much - that would be more esthetically pleasing but it isn't actually important I don't think. Clear out the TODOs here if it is working for you?